### PR TITLE
New Igel behavior + Corrupted Igel

### PR DESCRIPTION
**Pull request by Rusty-Box**
The current Igel behavior is rather annoying and not really intuitive. I propose changing it to the following:

When Igel is approached and they spot the player, they will quickly roll into a spiky ball and dash forward towards the player, unrolling after a couple seconds and continuing to walk as usual until they spot the player again. They will no longer turn around when shot with fire!

A **scripted** proof-of-concept video of said behavior:

https://github.com/SuperTux/supertux/assets/20615048/76ecf7d0-5bce-44e5-82d3-2a2204bc385e

---

### Corrupted Igel

As with most corrupted variants, Corrupted Igel behaves the same as normal Igel, however, they are much faster and if killed when rolled into a ball, they shall burst out a couple spiky roots projectiles (similar to how Roof Crystallos shatter into damaging pieces upon falling on the ground), which break upon touching the ground!

TODO:

- [x] Missing "rolling-start" + "rolling" sprites for Corrupted Igel
- [x] Coding new Igel behavior
- [x] Coding Corrupted Igel behavior

### CODE REVIEW

**Comment by MatusGuy** [comment]
😁😁😁

**Comment by MatusGuy** [comment]
Literally made in only one day

File: data/images/creatures/igel/corrupted/corrupted_igel.sprite [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    (name "right")
+    (hitbox 18 24 36 38)
+	(fps 15)
+    (mirror-action "left")
+  )
+  (action
+    (name "roll-start-left")
+    (hitbox 16 24 36 38)
+	(fps 20)
+	(loops 1)
```
Comment (Vankata453) [inline data/images/creatures/igel/corrupted/corrupted_igel.sprite:26]: Wrong tabulation here and a few other places.

File: src/badguy/igel.cpp [hunk 2]
```diff
@@ -16,101 +16,220 @@
...
 #include "badguy/igel.hpp"
 
-#include "object/bullet.hpp"
+#include <variant>
+
+#include "audio/sound_manager.hpp"
+#include "collision/collision_system.hpp"
+#include "object/player.hpp"
+#include "object/shard.hpp"
+#include "math/easing.hpp"
```
Comment (Vankata453) [inline src/badguy/igel.cpp:25]: Order the includes alphabetically.

File: src/badguy/igel.cpp [hunk 3]
```diff
@@ -16,101 +16,220 @@
...
+
+      if (!m_roll_cooldown.started() && should_roll()) roll();
+
+      break;
+  }
 }
 
-void
-Igel::turn_around()
+void Igel::collision_solid(const CollisionHit &hit)
```
Comment (Vankata453) [inline src/badguy/igel.cpp:122]: `void` should be on another line. Same for other cases.

File: src/badguy/igel.cpp [hunk 4]
```diff
@@ -16,101 +16,220 @@
...
-    BadGuy::active_update(dt_sec);
-    return;
-  }
+bool Igel::should_roll()
+{
+  using RaycastResult = CollisionSystem::RaycastResult;
 
-  // Otherwise, adhere to the default behavior for WalkingBadguy.
-  WalkingBadguy::active_update(dt_sec);
+  Player* plr = get_nearest_player();
```
Comment (Vankata453) [inline src/badguy/igel.cpp:181]: Name it `player` so it's clear.

File: src/badguy/igel.cpp [hunk 5]
```diff
@@ -16,101 +16,220 @@
...
+  if (!plr) return false;
+
+  Rectf pb = plr->get_bbox();
```
Comment (Vankata453) [inline src/badguy/igel.cpp:184]: `player_box`

File: src/badguy/igel.hpp [hunk 6]
```diff
@@ -19,34 +19,44 @@
 
 #include "badguy/walking_badguy.hpp"
 
-/** Badguy "Igel" - a hedgehog that can absorb bullets */
+#include "supertux/timer.hpp"
+
+/** Giggle */
```
Comment (Vankata453) [inline src/badguy/igel.hpp:24]: Add descriptive description.

**Review by Vankata453 — APPROVED** [review]

**Review by mrkubax10 — APPROVED** [review]

**Comment by VincentAraneo** [comment]
I remember seeing that sprite in the game files when I first installed the 0.7 nightly builds. But they never used that sprite, until you used it for that new igel behavior to the game. I think it's cool!

**Pull request merged**